services/vision/mlvision/ml_model.go

```diff
@@ -0,0 +1,210 @@
+			if err != nil {
+				return nil, err
+			}
+			return registerMLModelVisionService(ctx, c.ResourceName().Name, attrs, actualR, logger)
+		},
+	})
+}
+
+// MLModelConfig specifies the parameters needed to turn an ML model into a vision Model.
+type MLModelConfig struct {
+	resource.TriviallyValidateConfig
+	ModelName string `json:"mlmodel_name"`
+}
+
+func registerMLModelVisionService(
+	ctx context.Context,
+	name string,
+	params *MLModelConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (vision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerMLModelVisionService")
+	defer span.End()
+
+	mlm, err := mlmodel.FromRobot(r, params.ModelName)
+	if err != nil {
+		return nil, err
+	}
+	classifierFunc, err := attemptToBuildClassifier(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+	}
+	detectorFunc, err := attemptToBuildDetector(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a detector:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+	}
+	segmenter3DFunc, err := attemptToBuild3DSegmenter(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a 3D segmenter:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service 3D segmenter", params.ModelName)
+	}
+	// Don't return a close function, because you don't want to close the underlying ML service
+	return vision.NewService(name, r, nil, classifierFunc, detectorFunc, segmenter3DFunc)
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch v := me.(type) {
+	case []uint8:
+		for _, t := range v {
+			out = append(out, float64(t))
+		}
+	case []float32:
+		for _, t := range v {
+			out = append(out, float64(t))
+		}
+	}
+	return out
+}
+
+// getLabelsFromMetadata returns a slice of strings--the intended labels.
+func getLabelsFromMetadata(md mlmodel.MLMetadata) []string {
+	for _, o := range md.Outputs {
+		if !strings.Contains(o.Name, "category") && !strings.Contains(o.Name, "probability") {
+			continue
+		}
+
+		if labelPath, ok := o.Extra["labels"]; ok {
+			var labels []string
+			f, err := os.Open(labelPath.(string))
+			if err != nil {
+				return nil
+			}
+			defer func() {
+				if err := f.Close(); err != nil {
+					return
+				}
+			}()
+			scanner := bufio.NewScanner(f)
+			for scanner.Scan() {
+				labels = append(labels, scanner.Text())
+			}
+			// if the labels come out as one line, try splitting that line by spaces or commas to extract labels
+			if len(labels) == 1 {
+				labels = strings.Split(labels[0], ",")
+			}
+			if len(labels) == 1 {
```

Comment:
this is duped code

Reply:
Naw. I specifically need to check one after the other. Split by comma and if it's still one thing _then_ split by space.

Reply:
I think you can use SplitN then

Reply:
SplitN just maxes out the number of output substrings

Reply:
got it. maybe write a comment for that!